### ISSUE TITLE
Gunicorn sync workers close connections unexpectedly

### ISSUE THREAD

**Issue by sverdoux** [benoitc/gunicorn#2881]
We are encountering strange connections close using gunicorn sync workers and sequentially sending POST requests from one client. Maybe related to #2714 

Minimal example to reproduce:
- server_side (ubuntu 20.04, python 3.8.10)
```
Flask==2.1.2
gunicorn==20.1.0
gevent==22.8.0
```
server file wsgi.py
```py
# for gevent tests
# from gevent import monkey
# monkey.patch_all()

from flask import Flask

app = Flask(__name__)


@app.route('/', methods=["GET", "POST"])
def index():
    '''Index page route'''

    return '<h1>Application Deployed!</h1>'
```
```
gunicorn -b 0.0.0.0:9098 -t 300 --workers 4 --worker-class sync --access-logfile - --error-logfile - --preload --log-level debug wsgi:app
```
- client side (other ubuntu 20.04, no localhost as the issue is not (less ?) reproducible
```py
import requests
import logging

logging.basicConfig(level=logging.DEBUG)

url = "http://<host>:9098/"

sample_payload = {
  "test_object": {
    "a": 1,
    "b": "test_string"
  },
    "id": "00000000-0000-0000-0000-000000000000"
}

for ii in range(0, 10000):
    response = requests.request("POST", url, json=sample_payload)
```

A connection reset error is raised from the client after a random number of iterations (sometimes less than 10, sometimes 100 ...)
```
requests.exceptions.ChunkedEncodingError: ("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer'))
```

What I have noticed:
- doing the same test with GET requests does not seem to produce the error
- using gevent workers (uncomment monkey_patch) solve the issue (a log in gevent worker explicitely describe proper close connection)
- gthread worker seems more robust, but still triggers the error 

As mentioned in the related issue , I dont understand the link with the backlog as here, I also have only one client to the server with sequential requests.
We will probably go the 'gevent worker' depending on the impact on our application as well if we can't find an other solution.

**Comment by leonidumanskiy** [comment]
I can confirm this looks very similar to what I have observed in #2714 

Unfortunately I did also observe the issue with `monkey.patch_all()` as well.

@sverdoux Have you tried inserting a small delay/sleep between your requests to see if the issue goes away or gets better/worse? I have a suspicion that there is a small time window during which a subsequent request would fail.

**Comment by sverdoux** [comment]
Adding a small delay using time.sleep between requests does indeed seem to remove the issue. My understanding of what is going on in the socket layer is limited but I thought that sync workers would be easily DOS  by multiple requests in parallel. The documentation mention that sync workers close connections on response so in this example scenario I am not sure to understand how workers would be 'under pressure'.
The delay is a lead and a proper buffering proxy may also help I guess (as recommended). Here its a minimal case but in our production case the problem is slightly different: a badly configured client was redirected its POST request automatically because of badly formed URLs (HTTP 308). Python requests (and I guess most http client frameworks) is configured by default to forward redirections so 2 POST are sent in a small window without controlled delay. And it may be hard to ask clients to first control their redirection manually and second add a delay in between.
We will keep analysing our prod setup to see if we can remove the issue, just wanted to share this simple setup to check if something fishy is going on.

**Comment by benoitc** [comment]
no activity since awhile. closing feel free to create a new ticket if needed.
